crates/db_views_moderator/src/admin_purge_comment_view.rs

```diff
@@ -20,6 +20,7 @@ impl AdminPurgeCommentView {
     admin_person_id: Option<PersonId>,
     page: Option<i64>,
     limit: Option<i64>,
+    hide_mod_names: bool,
   ) -> Result<Vec<Self>, Error> {
     let mut query = admin_purge_comment::table
       .inner_join(person::table.on(admin_purge_comment::admin_person_id.eq(person::id)))
```

Comment:
This one might not be possible? Rust/Diesel seems to have an issue doing conditional joins; I can't construct a query that does inner join if hide_mod_names is false, and left join otherwise. It complains about type incompatibility and such, so I can't seem to get it to compile. If you can give me a working example, that would help a lot.

Reply:
example: 
![image](https://user-images.githubusercontent.com/4389156/175859887-b18225f7-d1ea-4565-9385-78bcabe1b252.png)

Reply:
https://github.com/LemmyNet/lemmy/blob/main/crates/db_views/src/post_view.rs#L52

Then make sure the select contains `.nullable()`

Reply:
this _almost_ works, but fails when !hide_mod_names:

![image](https://user-images.githubusercontent.com/4389156/176814230-c76958eb-c85f-45e1-8405-a8dfd3e6ed37.png)

basically, I need the query to be 'WHERE (mod_person_id = person.Id AND !hide_mod_names) OR (mod_person_id = admin_person_id_join)

Having trouble finding any documentation on the .and() and .or() methods or any similar :(

Reply:
Sorry, forgot to ping again @dessalines

Reply:
I don't understand why this wouldn't work then:

```rust
let admin_person_id_join = if !hide_modlog_names {
  if let Some(admin_person_id) = admin_person_id {
    admin_person_id
  } else { -1 }
} else { -1 };

...

.left_join(
  person::table.on(
    admin_purge_comment::admin_person_id
    .eq(person::id)
    .and(person::id.eq(admin_person_id_join))))
```

Make sure that admin_person_id_join is getting set correctly.

Reply:
In the case where hide_mod_names is disabled and no admin_person_id is specified, you would get nulls for Person.

Reply:
Hrm... maybe the `.and` portion of the query could be moved down to a filter then.

```rust
if show_mod_names {
  if let Some(admin_person_id) = admin_person_id {
    query = query.filter(person::id.eq(admin_person_id)
  } 
} else {
  query = query.filter(person::id.eq(-1));
}
```

Reply:
Your above one might work too, you just might need some parentheses.

Reply:
Unfortunately a filter would remove the entire entity (basically a where clause) which doesn't work. Unfortunately, I don't think Diesel is flexible enough to cleanly handle the join properly. I think its best to keep the post processing loop to remove the 'moderator' entity manually. Is that okay with you @dessalines ?

Reply:
Unfortunately a filter would remove the entire entity (basically a where clause) which doesn't work. Unfortunately, I don't think Diesel is flexible enough to cleanly handle the join properly. I think its best to keep the post processing loop to remove the 'moderator' entity manually. Is that okay with you @dessalines ?

Reply:
Try this:

```rust
let admin_person_id_join = admin_person_id.unwrap_or(PersonId(-1));
let show_mod_names = !hide_mod_names;
let show_mod_names_expr = show_mod_names.as_sql::<Bool>();

let admin_names_join = 
  admin_purge_comment::admin_person_id.eq(person::id)
  .and(
    show_mod_names_expr.or(person::id.eq(admin_person_id_join))
  );
  

let mut query = admin_purge_comment::table
  .inner_join(
    person::table
    .on(admin_names_join)
  )
  .inner_join(post::table)
  .select((
    admin_purge_comment::all_columns,
    Person::safe_columns_tuple(),
    post::all_columns,
  ))
  .into_boxed();
...
```